admin/preprints/views.py

```diff
@@ -81,6 +87,30 @@ def get_context_data(self, **kwargs):
         kwargs.update({'message': kwargs.get('message')})
         return super(PreprintView, self).get_context_data(**kwargs)
 
+    def update_subjects_for_provider(self, request, old_provider, new_provider):
+        subject_hierarchies = self.object.subject_hierarchy
+        new_subjects = []
+        subject_problems = 0
+        for hierarchy in subject_hierarchies:
+            subject = hierarchy[-1]
+            if old_provider._id == 'osf':
+                bepress_id = subject.id
+            else:
+                bepress_id = subject.bepress_subject_id
+            if new_provider._id == 'osf':
+                new_subject = new_provider.subjects.filter(id=bepress_id)
+            else:
+                new_subject = new_provider.subjects.filter(bepress_subject_id=bepress_id)
+            if not new_subject.exists():
+                subject_problems = subject_problems + 1
+                new_subject = subject
+            else:
+                new_subject = new_subject[0]
+            new_subjects.append(new_subject.hierarchy)
+        self.object.set_subjects(new_subjects, Auth(request.user))
+        if subject_problems:
+            messages.warning(request, 'Unable to find subjects in new provider for {} subject(s)'.format(subject_problems))
```

Comment:
It would be beneficial if you listed the subjects that weren't able to be matched. The number of problems is limited in its helpfulness because it also doesn't capture subjects that are higher in the subject hierarchy.